### PR TITLE
Adds a reader/writer into/from dictionaries

### LINKED ISSUE #91: Include a native Python dict `read` function
Since the most basic version of this package only requires the pyyaml dependency, there should be a basic `read` function that returns a native Python type.

There is currently a `read_to_list` function (that isn't exposed as a top-level function), that returns a nested list. But that isn't in the most developer friendly format because it has a mix of column headers and data in the list. A dictionary type feels more appropriate.

The dictionary could be structured with 2 keys, "columns" and "data", with "data" being a 2D nested list. Or it could have the columns names as keys and the values each being a list for that column.

### PR DESCRIPTION
**Pull request by dalonsoa**
This PR adds a second option using only builtin tools by loading tabular data into a dictionary. 

The reader gives the option to manually add the column names, reading them from a row in the file or just using some default values. 

Under the hood, they both use the existing reader/writer using lists. 

Close #91

### CODE REVIEW

File: csvy/readers.py [hunk 1]
```diff
@@ -299,3 +300,51 @@ def read_to_list(
...
     return data, header
+
+
+def read_to_dict(
+    filename: Path | str,
+    marker: str = "---",
+    encoding: str = "utf-8",
+    csv_options: dict[str, Any] | None = None,
+    yaml_options: dict[str, Any] | None = None,
+    *,
```
Comment (alexdewar) [inline csvy/readers.py:311]: What does this do? Does it mean `column_names` and `fillvalue` have to be keyword args?

File: csvy/writers.py [hunk 2]
```diff
@@ -310,6 +311,43 @@ def write_polars(
...
+    data: Any,
+    comment: str = "",
+    encoding: str = "utf-8",
+    *,
+    fillvalue: Any = None,
+    **kwargs: Any,
+) -> bool:
+    """Write the dictionary to the chosen file, adding it after the header.
+
+    It transofrms the dictionary into a tabular format before saving it using the
```
Comment (alexdewar) [inline csvy/writers.py:326]:
```suggestion
    It transforms the dictionary into a tabular format before saving it using the
```

File: csvy/writers.py [hunk 3]
```diff
@@ -310,6 +311,43 @@ def write_polars(
     return False
 
 
+@register_writer
+def write_dict(
+    filename: Path | str,
+    data: Any,
```
Comment (alexdewar) [inline csvy/writers.py:317]: How about:

```suggestion
    data: Mapping[str, Any],
```

File: tests/test_write.py [hunk 4]
```diff
@@ -126,6 +126,25 @@ class Writer:
...
+    from csvy.writers import write_dict
+
+    class Writer:
+        writerow = MagicMock()
+
+    mock_save.return_value = Writer
+    filename = tmpdir / "some_file.csv"
+
+    data = {"a": [1, 2, 3, 4], "b": [1, 2, 3], "c": [1, 2, 3, 4, 5]}
+    expected_rows = max(len(v) for v in data.values()) + 1  # +1 for the column names
```
Comment (alexdewar) [inline tests/test_write.py:141]: Not important, but you could write this as:

```suggestion
    expected_rows = max(map(len, data.values())) + 1  # +1 for the column names
```

**Review by alexdewar — APPROVED** [review]
Looks good! Nice clean code.

I've made a few small suggestions.

File: csvy/readers.py (hunk 1 above)
Comment (dalonsoa) [reply csvy/readers.py:311]: Yes. I've opted for this enforcement to clearly distinguish them from the other arguments, which are common for all readers.

File: csvy/writers.py (hunk 3 above)
Comment (dalonsoa) [reply csvy/writers.py:317]: The problem is that, then, I should type annotate all the other readers AND the registry with exactly what they ingest, which is not possible to do for arrays and dataframes unless the relevant optional dependencies are included. So, I rather annotate with `Any` to skip the static checking there and leave proper type checking at runtime.

Comment (alexdewar) [reply csvy/writers.py:317]: Ohh I see. Nvm then.

**Pull request merged**